### PR TITLE
LibGfx/JPEGWriter: Fix crash on macOS when csize coefficient is 0

### LINKED ISSUE #21108: LibGfx/JPEG: JPEGWriter fails when encoding empty canvas to jpeg image
In LibWeb with this HTML the JPEGEncoder is called to encode the canvas bitmap to a jpeg image:
```html
<script>
    const canvas = document.createElement('canvas');
    document.write(canvas.toDataURL('image/jpeg'));
</script>
```

The canvas image is fully transparent but I expected a totally black JPEG to be encoded:
image

But this assertion fails on macOS:
```
VERIFICATION FAILED: false at /Users/bplaat/Projects/serenity/Userland/Libraries/LibGfx/ImageFormats/JPEGWriterTables.h:78
0   liblagom-core.0.0.0.dylib           0x000000010350e210 ak_verification_failed + 136
1   liblagom-gfx.0.0.0.dylib            0x0000000103676634 Gfx::(anonymous namespace)::JPEGEncodingContext::encode_dc(Gfx::OutputHuffmanTable const&, short const*, unsigned char) + 200
2   liblagom-gfx.0.0.0.dylib            0x0000000103675b94 Gfx::(anonymous namespace)::JPEGEncodingContext::write_huffman_stream() + 196
3   liblagom-gfx.0.0.0.dylib            0x0000000103674aac Gfx::JPEGWriter::encode(AK::Stream&, Gfx::Bitmap const&, Gfx::JPEGEncoderOptions const&) + 1576
4   liblagom-web.0.0.0.dylib            0x000000010470d884 Web::HTML::serialize_bitmap(Gfx::Bitmap const&, AK::StringView, AK::Optional<double>) + 220
5   liblagom-web.0.0.0.dylib            0x000000010470d458 Web::HTML::HTMLCanvasElement::to_data_url(AK::StringView, AK::Optional<double>) + 144
6   liblagom-web.0.0.0.dylib            0x0000000104af75a8 Web::Bindings::HTMLCanvasElementPrototype::to_data_url(JS::VM&) + 580
7   liblagom-js.0.0.0.dylib             0x0000000103bfd17c JS::NativeFunction::call() + 104
8   liblagom-js.0.0.0.dylib             0x0000000103bfcd40 JS::NativeFunction::internal_call(JS::Value, JS::MarkedVector<JS::Value, 32ul>) + 400
9   liblagom-js.0.0.0.dylib             0x0000000103aeb214 JS::call_impl(JS::VM&, JS::FunctionObject&, JS::Value, AK::Optional<JS::MarkedVector<JS::Value, 32ul>>) + 996
10  liblagom-js.0.0.0.dylib             0x0000000103a7f650 JS::Bytecode::Op::Call::execute_impl(JS::Bytecode::Interpreter&) const + 3208
11  liblagom-js.0.0.0.dylib             0x0000000103a7096c JS::Bytecode::Interpreter::run_and_return_frame(JS::Realm&, JS::Bytecode::Executable&, JS::Bytecode::BasicBlock const*, JS::Bytecode::CallFrame*) + 1512
12  liblagom-js.0.0.0.dylib             0x0000000103a6fec4 JS::Bytecode::Interpreter::run(JS::Script&, JS::GCPtr<JS::Environment>) + 728
13  liblagom-web.0.0.0.dylib            0x00000001047abe1c Web::HTML::ClassicScript::run(Web::HTML::ClassicScript::RethrowErrors, JS::GCPtr<JS::Environment>) + 168
14  liblagom-web.0.0.0.dylib            0x0000000104748ca8 Web::HTML::HTMLScriptElement::execute_script() + 328
15  liblagom-web.0.0.0.dylib            0x0000000104749d24 Web::HTML::HTMLScriptElement::prepare_script() + 3876
16  liblagom-web.0.0.0.dylib            0x00000001047812a4 Web::HTML::HTMLParser::handle_text(Web::HTML::HTMLToken&) + 600
17  liblagom-web.0.0.0.dylib            0x0000000104777b54 Web::HTML::HTMLParser::run() + 368
18  liblagom-web.0.0.0.dylib            0x0000000104779778 Web::HTML::HTMLParser::run(AK::URL const&) + 124
19  liblagom-web.0.0.0.dylib            0x0000000104636968 Web::parse_document(Web::DOM::Document&, AK::Detail::ByteBuffer<32ul> const&) + 124
20  liblagom-web.0.0.0.dylib            0x000000010484e270 Web::FrameLoader::resource_did_load() + 1924
21  liblagom-web.0.0.0.dylib            0x000000010456a3f0 AK::Function<void (Web::CSS::CSSStyleRule const&)>::operator()(Web::CSS::CSSStyleRule const&) const + 84
22  liblagom-web.0.0.0.dylib            0x0000000104850400 Web::Resource::for_each_client(AK::Function<void (Web::ResourceClient&)>) + 240
23  liblagom-web.0.0.0.dylib            0x0000000104850f10 Web::Resource::did_load(AK::Badge<Web::ResourceLoader>, AK::Span<unsigned char const>, AK::HashMap<AK::DeprecatedString, AK::DeprecatedString, AK::CaseInsensitiveStringTraits, AK::Traits<AK::DeprecatedString>, false> const&, AK::Optional<unsigned int>) + 2324
24  liblagom-web.0.0.0.dylib            0x0000000104856874 AK::Function<void (AK::Span<unsigned char const>, AK::HashMap<AK::DeprecatedString, AK::DeprecatedString, AK::CaseInsensitiveStringTraits, AK::Traits<AK::DeprecatedString>, false> const&, AK::Optional<unsigned int>)>::CallableWrapper<Web::ResourceLoader::load_resource(Web::Resource::Type, Web::LoadRequest&)::$_0>::call(AK::Span<unsigned char const>, AK::HashMap<AK::DeprecatedString, AK::DeprecatedString, AK::CaseInsensitiveStringTraits, AK::Traits<AK::DeprecatedString>, false> const&, AK::Optional<unsigned int>) + 88
25  liblagom-web.0.0.0.dylib            0x0000000104856f78 AK::Function<void (AK::Span<unsigned char const>, AK::HashMap<AK::DeprecatedString, AK::DeprecatedString, AK::CaseInsensitiveStringTraits, AK::Traits<AK::DeprecatedString>, false> const&, AK::Optional<unsigned int>)>::operator()(AK::Span<unsigned char const>, AK::HashMap<AK::DeprecatedString, AK::DeprecatedString, AK::CaseInsensitiveStringTraits, AK::Traits<AK::DeprecatedString>, false> const&, AK::Optional<unsigned int>) const + 116
26  liblagom-web.0.0.0.dylib            0x0000000104857dc4 AK::Function<void (AK::ErrorOr<int, AK::Error>)>::CallableWrapper<Web::ResourceLoader::load(Web::LoadRequest&, AK::Function<void (AK::Span<unsigned char const>, AK::HashMap<AK::DeprecatedString, AK::DeprecatedString, AK::CaseInsensitiveStringTraits, AK::Traits<AK::DeprecatedString>, false> const&, AK::Optional<unsigned int>)>, AK::Function<void (AK::DeprecatedString const&, AK::Optional<unsigned int>)>, AK::Optional<unsigned int>, AK::Function<void ()>)::$_4>::call(AK::ErrorOr<int, AK::Error>) + 1748
27  WebContent                          0x0000000102f21258 AK::Function<void (AK::ErrorOr<int, AK::Error>)>::operator()(AK::ErrorOr<int, AK::Error>) const + 156
28  WebContent                          0x0000000102f20ed8 WebContent::ConnectionFromClient::handle_file_return(int, AK::Optional<IPC::File> const&, int) + 172
29  WebContent                          0x0000000102f229a4 WebContentServerStub::handle(IPC::Message const&) + 2132
30  liblagom-ipc.0.0.0.dylib            0x000000010312994c IPC::ConnectionBase::handle_messages() + 224
31  liblagom-core.0.0.0.dylib           0x00000001034e9920 AK::Function<void ()>::operator()() const + 84
32  liblagom-core.0.0.0.dylib           0x0000000103504aec Core::ThreadEventQueue::process() + 364
33  liblagom-core.0.0.0.dylib           0x00000001034e83ec Core::EventLoopImplementationUnix::exec() + 44
34  liblagom-core.0.0.0.dylib           0x00000001034e71bc Core::EventLoop::exec() + 80
35  WebContent                          0x0000000102f18ab0 serenity_main(Main::Arguments) + 4056
36  WebContent                          0x0000000102f87b88 main + 216
37  dyld                                0x000000018876bf28 start + 2236
```

Hopefully someone knows what is wrong :^)

### PR DESCRIPTION
**Pull request by warpdesign**
This fixes #21108

### CODE REVIEW

**Review by bplaat — CHANGES_REQUESTED** [review]
Can you uncomment this test: https://github.com/SerenityOS/serenity/blob/master/Tests/LibWeb/Text/input/canvas/export.html#L16

The file https://github.com/SerenityOS/serenity/blob/master/Tests/LibWeb/Text/expected/canvas/export.txt becomes then: 
```
1. "image/png"
2. "image/jpeg"
```

File: Userland/Libraries/LibGfx/ImageFormats/JPEGWriter.cpp [hunk 1]
```diff
@@ -326,7 +326,7 @@ class JPEGEncodingContext {
     static u8 csize(i16 coefficient)
     {
         VERIFY(coefficient >= -2047 && coefficient <= 2047);
-        return floor(log2(abs(coefficient))) + 1;
+        return (u8)floor(log2(abs(coefficient))) + 1;
```
Comment (gmta) [inline Userland/Libraries/LibGfx/ImageFormats/JPEGWriter.cpp:329]:
```suggestion
        return static_cast<u8>(floor(log2(abs(coefficient)))) + 1;
```

**Comment by LucasChollet** [comment]
If we go this way I would really like more explanations in the commit message. Like explicitly say that you're casting `-inf` to a `u8` (I'm so surprised that it's not UB 😅).

But anyway I don't think that it's a good way to solve that issue. The spec defines `csize` for a determined set of input and zero is not a part of it. IMO, if there is anything to do in this function is to `VERIFY` that the input is not null. There is definitely an issue upstream, let's not extend the spec to workaround one of our issues.

**Comment by warpdesign** [comment]
> If we go this way I would really like more explanations in the commit message. Like explicitly say that you're casting `-inf` to a `u8` (I'm so surprised that it's not UB 😅).
> 
> But anyway I don't think that it's a good way to solve that issue. The spec defines `csize` for a determined set of input and zero is not a part of it. IMO, if there is anything to do in this function is to `VERIFY` that the input is not null. There is definitely an issue upstream, let's not extend the spec to workaround one of our issues.

Problem is that coefficient is `0` so log2(abs(coefficient)) returns `-inf`.

In Linux & macOS we both have 0, difference is that the cast to `u8` appears to be needed on macOS, not on Linux, but both have the sale problem in that this function is called with `0` and that will yield to `-inf`.

**Comment by Hendiadyoin1** [comment]
Would a simple
```cpp
if (coefficient == 0)
    return 0;
```
Not do the job?

**Comment by warpdesign** [comment]
> Would a simple
> ```cpp
> if (coefficient == 0)
>     return 0;
> ```
> Not do the job?

I guess this would do it but according to @LucasChollet `coefficient` should not be 0 in the first place.

**Comment by LucasChollet** [comment]
I will check that in details this weekend, but IIRC AC coefficient with a value of 0 are encoded differently. So I'm quite doubtful that this patch solves the issue, it doesn't crash anymore but is the output a properly encoded JPEG?

Also, sharing the bitmap that causes the issue would be useful.

**Comment by warpdesign** [comment]
> I will check that in details this weekend, but IIRC AC coefficient with a value of 0 are encoded differently. So I'm quite doubtful that this patch solves the issue, it doesn't crash anymore but is the output a properly encoded JPEG?
> 
> Also, sharing the bitmap that causes the issue would be useful.

I updated my test page to use the generated jpeg dataUrl as src for an image element and it doesn't seem to be correct: it seems to be a valid jpeg file, but not the expected one.

Chrome output:

image


Ladybird output (macOS):

image

**Comment by warpdesign** [comment]
This patch (as suggested by @Hendiadyoin1) appears to correctly fix the problem and the generated jpeg seems to be correct:

> return coefficient == 0.0 ? 0 : floor(log2(abs(coefficient))) + 1;

**Comment by LucasChollet** [comment]
I thought that the issue was caused by AC coefficients, as `csize` is defined for AC coefficients in the spec. But the spec is not super clear on how to encode these DC parameters (you have to come up with the algorithm, only the output is described). And I ended up reusing this `csize` function, without checking the result for a null input.

So let's get your fix in! I think that I prefer Hendiadyoin's version though. Enough things are already going on over this line, no need to add a ternary operator. And no need to check for equality with `0.0`, `coefficient` is a `i16` so 0 is fine.

**Review by LucasChollet — APPROVED** [review]

**Comment by awesomekling** [comment]
This is failing the lint check. Please run clang-format before uploading code :)

**Comment by BertalanD** [comment]
Does the JPEG that's generated when the new early return is hit pass validation?

**Comment by warpdesign** [comment]
> Does the JPEG that's generated when the new early return is hit pass validation?

How can I check JPEG validation? I tried converting the base64 text to a file and it can be opened without any problem with image viewers I have tried.

**Review by bplaat — APPROVED** [review]
Looks good to me

**Pull request merged**